overlord/snapstate/handlers.go

```diff
@@ -1740,15 +1668,49 @@ func (m *SnapManager) startSnapServices(t *state.Task, _ *tomb.Tomb) error {
 	perfTimings := state.TimingsForTask(t)
 	defer perfTimings.Save(st)
 
-	_, snapst, err := snapSetupAndState(t)
+	snapsup, snapst, err := snapSetupAndState(t)
 	if err != nil {
 		return err
 	}
-
 	currentInfo, err := snapst.CurrentInfo()
 	if err != nil {
 		return err
 	}
+
+	// check if any previously disabled services are now no longer services and
+	// log messages about that
+	for _, svc := range snapst.LastActiveDisabledServices {
+		app, ok := currentInfo.Apps[svc]
+		if !ok {
+			logger.Noticef("previously disabled service %s no longer exists", svc)
+		} else if !app.IsService() {
+			logger.Noticef("previously disabled service %s is now an app and not a service", svc)
+		}
+	}
+
+	// get the services which should be disabled (not started),
+	// as well as the services which are not present in this revision, but were
+	// present and disabled in a previous one and as such should be kept inside
+	// snapst for persistent storage
+	svcsToDisable, svcsToSave, err := missingDisabledServices(snapst.LastActiveDisabledServices, currentInfo)
+	if err != nil {
+		return err
+	}
+
+	// append services that were disabled by hooks (they should not get re-enabled)
+	svcsToDisable = append(svcsToDisable, snapst.ServicesDisabledByHooks...)
+
+	// commit the missing services to state so when we unlink this revision and
+	// go to a different revision with potentially different service names, the
+	// currently missing service names will be re-disabled if they exist later
+	t.Set("old-last-active-disabled-services", snapst.LastActiveDisabledServices)
+	snapst.LastActiveDisabledServices = svcsToSave
```

Comment:
I'd prefer if you separated the `t.Set` from this line because the comment currently only really applies to the second part, something like this

```suggestion
	// save the current last-active-disabled-services before we re-write it in case we
	// need to undo this
	t.Set("old-last-active-disabled-services", snapst.LastActiveDisabledServices)

	// commit the missing services to state so when we unlink this revision and
	// go to a different revision with potentially different service names, the
	// currently missing service names will be re-disabled if they exist later
	snapst.LastActiveDisabledServices = svcsToSave
```